### DISCUSSION TITLE
Dedicated pre-built looper control layouts/banks

### DISCUSSION THREAD

**Discussion by simonaglover95** [Pirate-MIDI/Pirate-MIDI-Features-Bug-Tracking#50] [Feature Requests & Ideas]
Originally submitted on looped.in ideas page with 17 community votes

**Comment by simonaglover95** [comment]
The current Bridge firmware can be easily used to control stuff like half/full speed or forward/reverse on loopers. But not for proper looper transport control (Rec / Overdub / Play / Replay from Beginning / Stop), since this needs grouped switches and LEDs, and an internal counter for the current state of the looper.

Suggestion: Implement this as Bank Modes. Regular Bank Mode would be separate switches (just as the Bridge behaves right now). Then, there could be 2, 3 and maybe 4 Switch looper modes, where those switches have dedicated looper control functionality; while the other switches remain freely assignable.

Background:

Disaster Area has been building MIDI Switchers allowing control of loop recorders since a decade. With my 2012 Gen1 DMC3XL (on my pedal board since then), I can easily control the built-in looper of my Strymon Timeline; as demonstrated in this vid I made:
https://www.youtube.com/watch?v=6ZCnSJyG0y8&t=367s

On the DMC3XL, there’s on footswitch for Rec/Overdub and a second for Play/Stop (2 switch). On the Timeline itself, there’s Rec/Overdub, Play/Restart and Stop (3 switch). Both implementations require that at least the Rec/Overdub and Play switches “know” the current state of the looper (or at least the state the looper is supposed to be in, due to the MIDI commands sent so far).

Like, when you press Rec from STOP state, looper will go into recording.
Pressing Rec from REC or PLAY state, looper will go into overdub.
Pressing Rec From OVERDUB state, looper will go into play.
Pressing Play from PLAY state will either stop (2 Switch implementation) or restart the loop (3 Switch implementation).
Pressing Play from STOP, REC or OVERDUB state will sent the looper into PLAY mode.
All of this will result in the Rec and Play Switch LEDs lighting up in ways reflecting the current state (both on = OVERDUB). 

Not sure if this could be translated into a generic implementation of user-assignable grouping of switches and LEDs which would still be user-understandable. That’s why I’m rather suggesting specific looper modes above.

**Comment by ultracello** [comment]
Yes, with running Loopy Pro this would be super cool. +1 !
Top user idea!

**Reply by wpostma** [reply]
Very interested in seeing a Looper Integration Mode as a major mode for PirateMidi Bridge X controllers.

Loopy Pro,  various DAWs with scripting (Bitwig) would benefit from a looper state writeable via Midi Sends to the Bridge 6.

**Reply by ultracello** [reply]
Yeah! Btw. see also this here:

https://roadmap.loopypro.com/feature-requests/p/pirate-midi-bridge-146-open-api-controller-support

And as bonus also these two ones here:
– https://roadmap.loopypro.com/feature-requests/p/customizable-info-fields-auv3-plugin-parameter-values-shown-on-fader-and-other-w
– https://roadmap.loopypro.com/feature-requests/p/for-labels-on-widgets-placeholders-for-auv3-plugin-parameters

**Comment by wpostma** [comment]
I have been thinking about this a bit more and there's another case ; Integration with DAWs.

I have for example a Novation Launchpad, and other devices, all of which have custom SCRIPTS or extension modules (dlls, or java or javascript or LUA scripts) for various controllers, in the varoius formats that various DAWs use.

What these have in common is:

1.   The device in question, when used for controlling looping should have some STATE variables inside the product such as writeable locations using some MIDI writes (usually done as midi CC messages from the DAW host to the controller).   

2. There should be a writeable state for each button or switch on a controller.  IN the case of the PirateMidi bridge6, I would like a mode where the switch generates CCs (as it always does) but also knows that its LED state is remotely managed, and that whether the left LED or the right LED or both are lit, or are flashing or half or full bright, or whatever you can do in hardware, should be CONTROLLABLE from the DAW's controller scripts.

3.  I would like to have one button to be my loop record/play/overdub controller button and as the end user, I would want to know if it's idle,  playing or overdubbing. On a real physical looper pedal I might have LED color 1 for record and LED COLOR 2 for play, and flashing to count down when its about to start or about to stop.   This would required my DAW script to ahve control over the LED lights.

4.  Since I'm asking for LED LIGHT CONTROL it might also be nice to have control over what shows in the display (PLAY, RECORD, etc).  This would be accomplished by writing to a series of MIDI CC values (12 in a row) in order to set the 12 characters of scribble script text.

**Comment by TheGuacamoleXplosion** [comment]
The original post was made by me. With sequential switches, I was able to implement both a 2-switch looper transport control à la Disaster Area DMC3 and a 3-switch looper transport control similar to the one on board the Strymon Timeline. 

Here’s a vid: 
https://www.youtube.com/watch?v=csM9ZxwKrIM

There’s still two problems however:

1. State of Sequential Switches is not preserved when changing banks. So, when I have my looper play, record or overdub; then change banks; then come back to my looper bank — the Sequential Switches will not know their current state, and will not display the according LED colours and text labels. 
https://github.com/Pirate-MIDI/Pirate-MIDI-BridgeOS/issues/95

2. Sequential Switches currently always switch _on release_. This is not convenient for looper control: You want your RECORD, PLAY and STOP actions to happen immediately on press; getting correct timing when looping manually is hard enough _on press_, if you’re not a drummer or a organ footpedal player. 
So, it would be great to have optional “Sequential on press”/“Linked sequential on press” modes. On-press-switching of course is not compatible with having independent secondary hold functionality, so the old “Sequential on release” modes should not be abandoned.


See also description here: https://discord.com/channels/1026256363174764647/1026723004266324048/threads/1117904066601816094
